lib/srv/desktop/rdp/rdpclient/src/rdpdr.rs

```diff
@@ -72,30 +73,91 @@ pub struct TeleportRdpdrBackend {
     active_device_ids: Vec<u32>,
     /// The client handle for this backend, used to send messages to the RDP server.
     client_handle: ClientHandle,
+    // contexts holds all the active contexts for the server, established using
+    // SCARD_IOCTL_ESTABLISHCONTEXT. Some IOCTLs are context-specific and pass it as argument.
+    //
+    // contexts also holds a cache and connected smartcard handles for each context.
+    contexts: Contexts,
 }
 
 impl TeleportRdpdrBackend {
     pub fn new(smartcard_device_id: u32, client_handle: ClientHandle) -> Self {
         Self {
             active_device_ids: vec![smartcard_device_id],
             client_handle,
+            contexts: Contexts::new(),
         }
     }
 
     fn get_scard_device_id(&self) -> PduResult<u32> {
-        if self.active_device_ids.len() == 0 {
+        if self.active_device_ids.is_empty() {
             return Err(other_err!(
                 "TeleportRdpdrBackend::get_scard_device_id",
                 "no active devices",
             ));
         }
         Ok(self.active_device_ids[0])
     }
+
+    fn handle_access_started(
+        &mut self,
+        req: DeviceControlRequest<ScardIoCtlCode>,
+        _call: ScardAccessStartedEventCall,
+    ) -> PduResult<()> {
+        if req.header.device_id != self.get_scard_device_id()? {
+            return Err(other_err!(
+                "TeleportRdpdrBackend::handle_scard_access_started_event_call",
+                "got ScardAccessStartedEventCall for unknown device_id",
```

Comment:
Would it be helpful to include the expected and actual IDs in the error message?

Reply:
Yes, unfortunately `other_err` only allows us to input string literals (`&'static str`). I created a custom error type here which allows us to put arbitrary messages. I figure this gets the job done for now, and we can make this type more sophisticated as time goes on if needed.

5e2f7c66d50280e2d3c6362e5b04691583d97838